cognite/powerops/utils/preprocessor_utils.py

```diff
@@ -0,0 +1,174 @@
+"""
+Functions in this module are almost verbatim copies of functions in the `preprocessor.utils`
+from the power-ops-functions repo.
+"""
+# TODO refactor the power-ops-functions repo to use this module and get rid of duplicated code.
+
+import logging
+from collections.abc import Iterable
+from typing import Optional, cast
+
+import arrow
+import numpy as np
+import pandas as pd
+from cognite.client import CogniteClient
+from cognite.client.data_classes import DatapointsList
+
+from cognite.powerops.utils.require import require
+
+logger = logging.getLogger(__name__)
+
+
+def ms_to_datetime(timestamp: int):
+    """Milliseconds since Epoch to datetime."""
+    return arrow.get(timestamp).datetime.replace(tzinfo=None)  # TODO: take tz into account
+
+
+def arrow_to_ms(arrow: arrow.Arrow) -> int:
+    """Arrow datetime to milliseconds since Epoch."""
+    return int(arrow.float_timestamp * 1000)
+
+
+def _overlapping_keys(dict_: dict, other: dict) -> list[str]:
+    return list(set(dict_.keys()).intersection(other.keys()))
+
+
+def merge_dicts(*args: dict) -> dict:
+    merged: dict = {}
+    for dict_ in args:
+        assert isinstance(dict_, dict), f"Expected dict, got {type(dict_)}!"
+        overlap = _overlapping_keys(dict_, merged)
+        if overlap:
+            raise Exception(f"Key collision on '{overlap}' when merging dictionaries!")
+        merged = {**merged, **dict_}
+    return merged
+
+
+# TODO: consider exception
+def log_missing(expected: Iterable, actual: Iterable) -> None:
+    if missing := set(expected).difference(actual):
+        logger.warning(f"Missing: {', '.join(map(str, missing))}")
```

Comment:
This is a bit hidden, lets propagate the missing in the `validate` function in `resync` and print them there using the `echo(Missing, warning=True)`